Zend/zend_inheritance.c

```diff
@@ -173,72 +174,205 @@ static zend_always_inline zend_bool zend_iterable_compatibility_check(zend_arg_i
-			proto_class_name = zend_string_copy(proto->common.scope->parent->name);
-		} else if (class_name_len == sizeof("self")-1 && !strcasecmp(class_name, "self") && proto->common.scope) {
-			proto_class_name = zend_string_copy(proto->common.scope->name);
+			if (!zend_string_equals_ci(fe_class_name, proto_class_name)) {
+				if (fe->common.type == ZEND_USER_FUNCTION) {
+					zend_class_entry * fe_ce = zend_lookup_class(fe_class_name);
+					zend_class_entry * proto_ce = zend_lookup_class(proto_class_name);
+
+					if (fe_ce && proto_ce) {
+						code = _check_zce_variance(fe_ce, proto_ce, variance);
+					} else {
+						/* The -1 will still be considered "truthy". It means there
+						 * is not enough information at the moment, but there are
+						 * not any known errors either. */
+						code = -1;
+					}
+				} else {
+					code = 0;
+				}
+			}
+			zend_string_release(proto_class_name);
+		} else if (variance == COVARIANT) {
+			if (proto_type_code == IS_ITERABLE) {
+				zend_class_entry * fe_ce = zend_lookup_class(fe_class_name);
+				code = fe_ce ? instanceof_function(fe_ce, zend_ce_traversable) : -1;
+			} else if (proto_type_code != IS_OBJECT) {
+				code = 0;
+			}
 		} else {
-			zend_string_addref(proto_class_name);
+			code = 0;
 		}
 
-		if (fe_class_name != proto_class_name && strcasecmp(ZSTR_VAL(fe_class_name), ZSTR_VAL(proto_class_name)) != 0) {
-			if (fe->common.type != ZEND_USER_FUNCTION) {
+		zend_string_release(fe_class_name);
+		return code;
+	} else if (ZEND_TYPE_IS_CLASS(proto_type)) {
+		if (variance == CONTRAVARIANT) {
+			if (fe_type_code == IS_ITERABLE) {
+				zend_string *proto_class_name =
+					_resolve_parent_and_self(proto, ZEND_TYPE_NAME(proto_type));
+				zend_class_entry *proto_ce = zend_lookup_class(proto_class_name);
 				zend_string_release(proto_class_name);
-				zend_string_release(fe_class_name);
-				return 0;
-			} else {
-				zend_class_entry *fe_ce, *proto_ce;
-
-				fe_ce = zend_lookup_class(fe_class_name);
-				proto_ce = zend_lookup_class(proto_class_name);
-
-				/* Check for class alias */
-				if (!fe_ce || !proto_ce ||
-						fe_ce->type == ZEND_INTERNAL_CLASS ||
-						proto_ce->type == ZEND_INTERNAL_CLASS ||
-						fe_ce != proto_ce) {
-					zend_string_release(proto_class_name);
-					zend_string_release(fe_class_name);
-					return 0;
-				}
+				return proto_ce
+					? instanceof_function(proto_ce, zend_ce_traversable)
+					: -1;
+			} else if (fe_type_code == IS_OBJECT) {
+				return 1;
 			}
 		}
-		zend_string_release(proto_class_name);
-		zend_string_release(fe_class_name);
-	} else if (ZEND_TYPE_CODE(fe_arg_info->type) != ZEND_TYPE_CODE(proto_arg_info->type)) {
-		/* Incompatible built-in types */
+	} else if (fe_type_code == IS_ITERABLE || proto_type_code == IS_ITERABLE) {
+		return (variance == COVARIANT && fe_type_code == IS_ARRAY)
+			|| (variance == CONTRAVARIANT && proto_type_code == IS_ARRAY);
+	} else if (fe_type_code == proto_type_code) {
+		return 1;
+	}
+
+	return 0;
```

Comment:
I think this code can be simplified and some of the duplication removed by performing some normalization. Doing a contravariant check of a, b should be the same as doing a covariant check of b, a, right?

Reply:
This is a very good insight; this should clean up a lot.